File: 구명보트.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

int solution(vector<int> people, int limit) {
	int answer = 0;
	sort(people.begin(), people.end(), greater<int>());
 
	int i = 0;
	int j = people.size() - 1;
	int sum = people[i] + people[j];

	// 100 80 70 50 40 
	while (i< j) {

		if (sum > limit) {
			answer++;
			i++;
			sum = people[i] + people[j];
		}
		else {
			j--;
			sum += people[j];
		}

	}
	if (i == j)answer++;

	
	
	return answer;
}

void main() {
	vector<int> people = { 100,70,80,50,40 };
	int limit = 200;
	cout<<solution(people, limit);
}

File: 프로그래머스/level2/구명보트.js
function solution (people, limit) {
    var answer = 0
    people.sort((a, b) => b - a)
    let l = 0
    let r = people.length - 1

    while (l < r) {
        var sum = people[l] + people[r]
        if (sum > limit) { //한 명만 탑승
            l++
        } else { // 두명 탑승 
            l++
            r--
        }
        answer++ //보트 추가 
    }
    if (l == r) answer++ // 마지막에 한명 남을 경우 보트 추가
    return answer
}

console.log(solution([70, 50, 80, 50], 100));